### DISCUSSION TITLE
[Panache & Panache Reactive] How is the support for optional parameters in a query?

### DISCUSSION THREAD

**Discussion by agreedSkiing** [quarkusio/quarkus#31875] [Q&A] [answered]
Can Panache somehow handle optional parameters without the use of the criteria query api or using something like `AND ( :parameter IS NULL OR c.column = :parameter )`? This would be parameters that might not exists depending on how the HQL query was built as an example from our up and coming rest api we will have multiple filters that look like `filter[X]` where `X` could be regions, municipalities, organization-numbers, ids, types, matchAll (is a boolean filter), categories and some more so just this would create a lot of different variations query strings and could be solved with criteria queries, probably, but the step from going to criteria queries to something that looks like SQL isn't the easiest.

Our project is currently using Hibernate Reactive and not Panache Reactive but I'm looking for alternatives since we made the switch from Panache Reactive to Hibernate Reactive once before and could probably do the switch back. 

P.S. My first guess (this is what I gathered from the Quarkus guide) is that Panache has the same support as Hibernate... so no special case for non existent parameters in the query when a Map of parameters or the Parameter object is added to the HQL statement.

[a comment by an automated account is not shown]

**Reply by loicmathieu** [reply, marked as answer]
As you said, Panache has the same support than Hibernate. 
There is open issue about this, you can add feedback and +1 them:

- Implements Specification or Example for Panache: https://github.com/quarkusio/quarkus/issues/8136
- Support QueryDSL in Panache: https://github.com/quarkusio/quarkus/issues/4016
- Jpa criteria support in Hibernate ORM Panache: https://github.com/quarkusio/quarkus/issues/3965
- Add support for optional predicates in Panache: https://github.com/quarkusio/quarkus/issues/2303

There is also other extensions that provides this functionality, as Panache entities are JPA entities they can be used everywhere JPA can be used. For example JPAStreamer: https://quarkiverse.github.io/quarkiverse-docs/quarkus-jpastreamer/dev/index.html

**Reply by agreedSkiing** [reply]
Ah so there are some proposals in the works for Panache, and quarkus-jpastreamer looks interesting since it works with Panache but I would probably go with Blaze persistence instead, but one thing I get the feeling with quarkus-jpastreamer which is also a problem with Blaze persistence is that it does not seem to support Hibernate Reactive that is used by [Panache Reactive](https://quarkus.io/guides/hibernate-reactive-panache) and the current [Hibernate Reactive](https://quarkus.io/guides/hibernate-reactive)

**Reply by lucamolteni** [reply]
@agreedSkiing this is a pretty old question but I received a notification so I think it might be worth telling you that Quarkus Data is out https://quarkus.io/blog/introducing-quarkus-data/ and you're a power user of both Panache and Hibernate Reactive so it might be worth taking a look at it

**Reply by agreedSkiing** [reply]
That was a good read :), thanks @lucamolteni.
